python/lancedb/embeddings/functions.py

```diff
@@ -0,0 +1,183 @@
+        """
+        This creates a decorator that can be used to register
+        an EmbeddingFunctionModel.
+        """
+
+        # This is a decorator for a class that inherits from BaseModel
+        # It adds the class to the registry
+        def decorator(cls):
+            if not issubclass(cls, EmbeddingFunctionModel):
+                raise TypeError("Must be a subclass of EmbeddingFunctionModel")
+            if cls.__name__ in self._functions:
+                raise KeyError(f"{cls.__name__} was already registered")
+            self._functions[cls.__name__] = cls
+            return cls
+
+        return decorator
+
+    def reset(self):
+        """
+        Reset the registry to its initial state
+        """
+        self._functions = {}
+
+    def load(self, name: str):
+        """
+        Fetch an embedding function class by name
+        """
+        return self._functions[name]
+
+    def parse_functions(self, metadata: dict):
+        """
+        Parse the metadata from an arrow table and
+        return a mapping of the vector column to the
+        embedding function and source column
+
+        Parameters
+        ----------
+        metadata : dict
+            The metadata from an arrow table. Note that
+            the keys and values are bytes
+        """
+        serialized = metadata[b"embedding_functions"]
+        raw_list = json.loads(serialized.decode("utf-8"))
+        functions = {}
+        for obj in raw_list:
+            model = self.load(obj["schema"]["title"])
+            functions[obj["model"]["vector_column"]] = model(**obj["model"])
+        return functions
+
+    def function_to_metadata(self, func):
+        """
+        Convert the given embedding function and source / vector column configs
+        into a config dictionary that can be serialized into arrow metadata
+        """
+        schema = func.model_json_schema()
+        json_data = func.model_dump()
+        return {
+            "schema": schema,
+            "model": json_data,
+        }
+
+    def get_table_metadata(self, func_list):
+        """
+        Convert a list of embedding functions and source / vector column configs
+        into a config dictionary that can be serialized into arrow metadata
+        """
+        json_data = [self.function_to_metadata(func) for func in func_list]
+        # Note that metadata dictionary values must be bytes so we need to json dump then utf8 encode
+        metadata = json.dumps(json_data, indent=2).encode("utf-8")
+        return {"embedding_functions": metadata}
+
+
+REGISTRY = EmbeddingFunctionRegistry()
+
+
+class EmbeddingFunctionModel(BaseModel, ABC):
+    """
+    A callable ABC for embedding functions
+    """
+
+    source_column: Optional[str]
+    vector_column: str
+
+    @abstractmethod
+    def __call__(self, *args, **kwargs):
+        pass
+
+
+@REGISTRY.register()
+class SentenceTransformerEmbeddingFunction(EmbeddingFunctionModel):
+    name: str = "all-MiniLM-L6-v2"
```

Comment:
This is an example that is registered by default?

Reply:
This is the first of a set of "default" functions. The name variable is the model name ; sentence-transformers can be configured with a list of different models.

In a follow on PR we will add OpenAI and OpenClip embedding functions as well.